ToolChain/Assembler/Assembler.py

```diff
@@ -56,79 +50,90 @@ def __init__(self, inputFile=None, outputFile=None):
+        Internal and external symbol lists contain tuples containing the symbol name, and its offset.
+        These will be used to fill the .o file information before the binary data.
+        :param inputFile: str, the name of the file to be read.
+        :param outputFile: str, the name of the file to be written to.
         :return:
         """
 
-        file = open(file, mode="wb")
-        file.write(outputData)
+        parser = Parser()                    # Parser object which will return the binary code for instructions
+        masterString = b""                   # String that will eventually be our output file
+        xmlstring = b""                      # String that will contain the "xml" data at the beginning of the .o file
+        globalList = []                      # Temporary list to hold the names of global(external) symbols
+        externalList = []                    # Master list of all external symbols to be used by linker
+        internalList = []                    # Master list of all internal symbols
+        offset = 0                           # Offset used to calculate a label's offset from the beginning of the file
+        lineno = 0                           # Keeps track of which line we're parsing to pinpoint errors
+
+        instructionFlag = 0                  # Flag for an operand, meaning parser found a valid instruction
+        localReferenceFlag = 1               # Flag for local function, internal symbol
+        globalReferenceFlag = 2              # Flag for external function, global symbol
+        commentFlag = 3                      # Flag for comment
+
+        file = open(inputFile, mode="r")
+        fileLines = file.readlines()
         file.close()
+
+        masterString += b"<Text>"
+
+        for line in fileLines:
+            lineno += 1
+            # If the file isn't empty, process the content
+
+
+            instruction, offset, labelFlag = parser.parse(line)
+
+            if labelFlag == instructionFlag:
+                masterString += instruction
+
+            elif labelFlag == localReferenceFlag:
+                labelTuple = (instruction, offset)
+                internalList.append(labelTuple)
+                if instruction in globalList:
+                    externalList.append(labelTuple)
+
+            elif labelFlag == globalReferenceFlag:
+                globalList.append(instruction)
+
+            elif labelFlag == commentFlag:
+                # flag for comment, we simply ignore and move on to the next line
+                pass
+
+            else:
+                raise ValueError("Couldn't parse instruction at line " + lineno)
+
+        masterString += b"</Text>"
+
+        # builds the output file by extracting and printing each symbol into its appropriate category.
+        xmlstring = b"<AssemblySize>" + struct.pack(">I", parser.relativeAddressCounter) + \
```

Comment:
I don't think the counter should be in the parser. The presence of the counter in the parser goes against OOP good practices and gives an additional job to the parser (tracking the assembly size). This should be the job of the Assembler class (enabled by the parser). This also "decouples" the Assembler and Parser classes which is a good thing for future improvements.